File: Baekjoon/Class3/19.1260 DFS와 BFS.java
import java.io.*;
import java.util.*;

class Main {
    private static StringBuilder sb;
    private static ArrayList<Integer>[] graph;
    private static boolean[] visited;

    // 깊이 우선 탐색
    private static void dfsRecursive(int start) {
        visited[start] = true;
        sb.append(start).append(' ');
        for(int end: graph[start])
            if(!visited[end]) {
                dfsRecursive(end);
            }
    }
    private static void dfs(int start) {
        visited = new boolean[graph.length];
        dfsRecursive(start);
        sb.append('\n');
    }

    // 너비 우선 탐색
    private static void bfs(int start) {
        Deque<Integer> deque = new ArrayDeque<>();
        visited = new boolean[graph.length];
        deque.addLast(start);
        visited[start] = true;
        sb.append(start).append(' ');
        while(!deque.isEmpty()) {
            start = deque.removeFirst();
            for(int end: graph[start])
                if(!visited[end]) {
                    deque.addLast(end);
                    visited[end] = true;
                    sb.append(end).append(' ');
                }
        }
        sb.append('\n');
    }

    public static void main(String[] args) throws IOException {
        int n, m, v;
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        sb = new StringBuilder();

        n = Integer.parseInt(st.nextToken());
        m = Integer.parseInt(st.nextToken());
        v = Integer.parseInt(st.nextToken());

        graph = new ArrayList[n + 1]; // 무방향성 그래프
        for(int i = 1; i <= n; i++)
            graph[i] = new ArrayList<>();
        for(int i = 0; i < m; i++) {
            st = new StringTokenizer(br.readLine());
            int start = Integer.parseInt(st.nextToken());
            int end = Integer.parseInt(st.nextToken());
            graph[start].add(end);
            graph[end].add(start);
        }
        for(int i = 1; i <= n; i++) // 정렬
            Collections.sort(graph[i]);

        dfs(v);
        bfs(v);

        System.out.print(sb);
        br.close();
    }
}

File: Baekjoon/Class3/19.1260 DFS와 BFS.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <queue>
using namespace std;

vector<vector<int>> graph;
vector<char> visited;

// 깊이 우선 탐색
void dfs_recursive(int start) {
    visited[start] = true;
    cout << start << ' ';
    for (int end : graph[start])
        if (!visited[end])
            dfs_recursive(end);
}

void dfs(int start) {
    visited = vector<char>(graph.size());
    dfs_recursive(start);
    cout << '\n';
}

// 너비 우선 탐색
void bfs(int start) {
    queue<int> q;
    visited = vector<char>(graph.size());
    q.push(start);
    visited[start] = true;
    cout << start << ' ';
    while (!q.empty()) {
        start = q.front();
        q.pop();
        for (int end : graph[start])
            if (!visited[end]) {
                q.push(end);
                visited[end] = true;
                cout << end << ' ';
            }
    }
    cout << '\n';
}

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);

    int n, m, v;
    cin >> n >> m >> v;

    graph = vector<vector<int>>(n + 1); // 무방향성 그래프
    while (m--) {
        int start, end;
        cin >> start >> end;
        graph[start].push_back(end);
        graph[end].push_back(start);
    }
    for (vector<int>& node : graph) // 정렬
        sort(node.begin(), node.end());

    dfs(v);
    bfs(v);

    return 0;
}